Support set the callback function type according to the event type.

zjffun opened the pull request:
### This PR implements or fixes... _(explain your changes)_

Add missing types.

Set callback events type to typescript conditional types, which can get the type of event according to the event name given before. 

### This PR closes the following issues... _(if applicable)_

#382 

### Does this PR require the Docs to be updated?

No

### Does this PR require new tests?

No

### This branch been tested on... _(click all that apply / add new items)_

**Browsers:**

* [ ] Chrome _version_
* [ ] Firefox _version_
* [ ] Safari _version_
* [ ] IE / Edge _version_
* [ ] iOS Browser _version_
* [ ] Android Browser _version_

zjffun commented:
Reproduction: 

```js
import { Sortable, SortableSortEvent } from '@shopify/draggable';

const sortable = new Sortable(document.querySelectorAll('ul'), {
  draggable: 'li'
});

function sort(evt: SortableSortEvent) {
  console.log(evt.dragEvent);
}

sortable.on('sortable:sort', sort);
sortable.off('sortable:sort', sort);
```

bahung1221 commented:
Hi @zjffun , I tested this PR, it worked very well!

But it seems we missing types for `Collidable` and `Snappable` Plugins, so I tried to add it in https://github.com/bahung1221/draggable/commit/d50f134cd9e49b28ba4d74c02acbf861b62cfd83 .

But when I add the above events types, the declaration of **Draggable** class looks too long, but I don't know another way to do it.
```
export class Draggable<EventListType extends string = DraggableEventNames | MirrorEventNames | CollidableEventNames | SnappableEventNames> {
  ...
}
```

Could you please help to check it? It still fine if you think we don't need to add them 😄

zjffun commented:
Thanks! The missing types are merged. I think here we can manually import the types of non-default plugins.

I will add how to using draggable with ts to the document.

```js
import {Droppable, Plugins} from '@shopify/draggable';
import type {DroppableEventNames, CollidableEventNames, CollidableEvent} from '@shopify/draggable';

const droppable = new Droppable<DroppableEventNames | CollidableEventNames>(document.querySelectorAll('.container'), {
  draggable: '.item',
  dropzone: '.dropzone',
  collidables: '.other-list',
  plugins: [Plugins.Collidable]
});

function collidable(evt: CollidableEvent){
  evt.canceled
}

droppable.on('droppable:dropped', () => console.log('droppable:dropped'));
droppable.on('droppable:returned', () => console.log('droppable:returned'));
droppable.on("collidable:in", collidable)
```

bahung1221 commented:
@tsov , could you please help to merge this PR?

BTW, we also have 4 PR that was approved, if you have free time, please help to check and merge it 😄 
Thank you so much.

tsov commented:
This just needs a rebase 🙂

zjffun commented:
@tsov This PR is rebased. Thank you.

tsov commented:
👌 Maybe good time for another release?

zjffun commented:
I will open a PR for beta11.
